homework_solution.md

```diff
@@ -0,0 +1,474 @@
+
+      ```
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id
+      UNION
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id;
+      ```
+
+   4. Insert into the `stations` table the result of our query
+
+      ```
+      INSERT INTO stations (
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+         UNION
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+      );
+      ```
+
+   5. Adjust the `id` sequence to be the highest `id` from our `INSERT` above
+
+      ```
+      SELECT setval(pg_get_serial_sequence('stations', 'id'), (SELECT MAX(id) FROM stations));
+      ```
+
+   6. Finally, add entries for records where a name exists, but the id is null, relying on the seqquence to generate an id. This has to be done in 2 steps, so we can exclude the records from the first insertions when we run the second one, avoiding duplicates.
+
+      ```
+      INSERT INTO stations (
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), from_station_name
+         FROM trips
+         WHERE
+            from_station_id IS NULL
+         AND
+            from_station_name NOT IN (SELECT name from stations)
+         GROUP BY from_station_name
+      );
+      INSERT INTO stations(
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), to_station_name
+         FROM trips
+         WHERE
+            to_station_id IS NULL
+         AND
+            to_station_name NOT IN (SELECT name from stations)
+         GROUP BY to_station_name
+      );
+      ```
+
+2. _**Should we add any indexes to the stations table, why or why not?**_
+
+   Not right now. We already have an index on `id`, because it's the table's pkey. For our current usage, that should suffice.
+
+3. _**Fill in the missing data in the `trips` table based on the work you did above**_
+
+   ### Steps
+
+   1. We _could_ update the names in `trips`, like so:
+
+      ```
+      UPDATE trips t
+      SET
+         from_station_name = (SELECT name FROM stations s WHERE t.from_station_id = s.id),
+         to_station_name = (SELECT name FROM stations s WHERE t.from_station_id = s.id)
+      ```
+
+   2. However, it's best to just delete the station names column, so the `stations` table becomes the single source of truth for that (but we won't do that just yet!!)
+
+      ```
+      ALTER TABLE trips_backup
+      DROP COLUMN from_station_name,
+      DROP COLUMN to_station_name;
+      ```
```

Comment:
[comment]
Great! The station names have now become denormalized data (and inconsistently as well), So it's smart to make the stations table the single source of truth.

What are the downsides to this approach, if any?

Reply:
The two (possible) downsides I can think of are:
1. Losing some of the original data. Now you can't tell just from looking at the `trips` table what the original values were
2. Depending on how this grows, joining the tables in order to get the station names could become costly.

Reply:
Very good!